src/lsm/scan_tree.zig

```diff
@@ -568,92 +605,109 @@ fn ScanTreeLevelType(comptime ScanTree: type, comptime Storage: type) type {
-                } else {
-                    self.manifest = .finished;
-                    break :blk null;
-                }
-            };
+                    break :table_next .{ .key_exclusive = key_last };
+                };
 
-            assert(self.iterator.callback == .none);
-            self.iterator.start(.{
-                .grid = scan.tree.grid,
-                .level = self.level_index,
-                .snapshot = scan.snapshot,
-                .index_block = self.buffer.index_block,
-                .tables = .{ .scan = table_info },
-                .direction = scan.direction,
-            });
+                self.state = .{
+                    .loading_index = .{
+                        .manifest_table_next = manifest_table_next,
+                        .address = table_info.address,
+                        .checksum = table_info.checksum,
+                    },
+                };
+                self.values = .fetching;
+            } else {
+                self.state = .{ .finished = .{} };
+                self.values = .finished;
+            }
         }
 
-        /// Fetches data from storage for the current `table_info`.
         pub fn fetch(self: *ScanTreeLevel) void {
-            assert(self.manifest == .iterating_blocks or
-                self.manifest == .finished);
-            assert(self.values == .fetching);
             assert(self.scan.state == .buffering);
-            assert(self.iterator.callback == .none);
 
-            self.iterator.next(.{
-                .on_index = index_block_callback,
-                .on_data = data_block_callback,
-            });
+            switch (self.state) {
+                .iterating_manifest => unreachable,
+                .loading_index => |*loading_index| {
+                    assert(self.values == .fetching);
+                    // Reading the index blocks:
+                    const grid: *Grid = self.scan.tree.grid;
+                    grid.read_block(
+                        .{ .from_local_or_global_storage = index_block_callback },
+                        &loading_index.read,
+                        loading_index.address,
+                        loading_index.checksum,
+                        .{ .cache_read = true, .cache_write = true },
+                    );
+                },
+                .iterating_values => |*iterating_values| {
+                    assert(self.values == .fetching);
+                    iterating_values.iterator.next(value_block_callback);
+                },
+                .finished => |*finished| {
+                    assert(self.values == .finished);
+                    const grid: *Grid = self.scan.tree.grid;
+                    grid.on_next_tick(
+                        finished_callback,
+                        &finished.next_tick,
+                    );
+                },
+            }
         }
 
         pub fn peek(self: *const ScanTreeLevel) error{ Drained, Empty }!Key {
```

Comment:
Out out scope of this PR, but one thing I am finding slightly unsatisfactory about the scans is that the implementation is very scalar, it doesn't batch. 

So for example, kway merge operates on a value at a time: in the inner loop for the kway merge there's a break edge for each iteration.

At the same time, our data is fundmentally batched --- a value block contains a whole bunch of values. Table memory is also an array. 

It seems like it would be more in-line with our batching philosophy to do a batching merge, pushing fors down:


```
fn kmerge(
    inputs: [][]const Value,
    output: []Value
) struct {
   values_written: usize, 
   input_exhausted: usize, // index into inputs
}
```

That is, `kmerge` takes entire _slices_, and, in the inner loop, merges slices until one of them is exhausted. The exhausted slice is communicated by the caller (the amount of values consumed is communicated by mutating `inputs`). The caller then does some asynchronous work to re-fill the exhausted batch, and then re-runs `kmerge`. 

One obstacle here is that our inputs are not acteally slices. Or, rather, they are slices for levels, but they have some duplicates for table_memory case. But it _should_ be possible to dedup when we sorting, right?

Reply:
Ahhhh, right! I was thinking that the `output` would be just the final message where we are copying the value into (this would also handle the limit case, as the merge could stop if the _output_ is exhausted as well), but, indeed, when we combine several scans that doesn't work.